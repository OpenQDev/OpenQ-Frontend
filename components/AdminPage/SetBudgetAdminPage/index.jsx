import React, { useContext, useEffect, useState } from 'react';
import useWeb3 from '../../../hooks/useWeb3';
import StoreContext from '../../../store/Store/StoreContext';
import useIsOnCorrectNetwork from '../../../hooks/useIsOnCorrectNetwork';
import TokenFundBox from '../../TokenSelection/TokenFundBox';
import ConnectButton from '../../WalletConnect/ConnectButton';
import AdminModal from '../AdminModal';
import TokenContext from '../../TokenSelection/TokenStore/TokenContext';

const SetBudgetAdminPage = ({ refreshBounty, bounty }) => {
  const [appState] = useContext(StoreContext);

  const [tokenState] = useContext(TokenContext);
  const { token } = tokenState;
  const [volume, setVolume] = useState('');
  const { openQClient, accountData, utils, logger } = appState;
  const [modal, setModal] = useState({});
  const [isOnCorrectNetwork] = useIsOnCorrectNetwork();
  const { account, library } = useWeb3();

  const [, tokenDispatch] = useContext(TokenContext);
  useEffect(() => {
    const depositTokenAddress = bounty?.deposits[0]?.tokenAddress;
    const payoutTokenAddress = bounty?.payoutTokenAddress;
    if (bounty?.bountyType == '1' && bounty?.deposits?.length > 0) {
      const tokenAddressDispatch = {
        type: 'SET_TOKEN',
        payload: {
          ...appState.tokenClient.getToken(depositTokenAddress),
          address: depositTokenAddress,
        },
      };
      tokenDispatch(tokenAddressDispatch);
    } else if (bounty?.bountyType == '1' && bounty?.payoutTokenVolume > 0) {
      const tokenAddressDispatch = {
        type: 'SET_TOKEN',
        payload: {
          ...appState.tokenClient.getToken(payoutTokenAddress),
          address: payoutTokenAddress,
        },
      };
      tokenDispatch(tokenAddressDispatch);
    }
  }, [bounty]);

  async function setBudget() {
    setModal({ type: 'Loading', inProgress: 'Updating Budget...' });
    try {
      const transaction = await openQClient.setFundingGoal(library, bounty.bountyId, token, volume);
      refreshBounty();
      setVolume('');
      setModal({
        transaction,
        type: 'Budget',
      });
    } catch (error) {
      logger.error(error, accountData.id, 'adminPage1');
      const { message, title } = openQClient.handleError(error, {
        bounty,
      });
      setModal({
        message,
        title,
        type: 'Error',
      });
    }
  }

  function onVolumeChange(volume) {
    utils.updateVolume(volume, setVolume);
  }

  return (
    <>
      {bounty.bountyType !== '3' && (
        <>
          {' '}
          <div className='flex items-center gap-2'>Set a New Budget for this Contract</div>
          <div className='flex-1 items-center w-full mt-2'>
            <TokenFundBox onVolumeChange={onVolumeChange} volume={volume} bounty={bounty} />
          </div>
          <ConnectButton nav={false} needsGithub={false} centerStyles={true} tooltipAction={'set a new budget.'} />
          {isOnCorrectNetwork && account && (
            <button className='btn-default' type='button' onClick={setBudget}>
              Set New Budget
            </button>
          )}
        </>
      )}

      <AdminModal setModal={setModal} tokenAddress={token.address} bounty={bounty} modal={modal} />
    </>
  );
};
export default SetBudgetAdminPage;
