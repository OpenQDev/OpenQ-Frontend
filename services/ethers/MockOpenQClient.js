/* eslint-disable */
// import { issueIds } from './mocks/data/issueIds';
import jsonRpcErrors from './JsonRPCErrors';
import axios from 'axios';
import { ethers } from 'ethers';

class MockOpenQClient {
    shouldSleep = 1000;
    shouldError = false;

    setSleep(time) {
        this.shouldSleep = time;
    }

    constructor(mockMutations) {
        
        const initialMockMutations = {
            setFundingGoal: () => { },
            setPayout: () => { },
            setSupportingDocumentsComplete: () => { },
            setTier: () => { },
            closeOngoing: () => { },
        }

        this.mockMutations = { ...initialMockMutations,...mockMutations,}; }

    async sleep(time = this.shouldSleep) {
        return new Promise(async (resolve) => {
            return setTimeout(resolve, time);
        });
    }

    reset() {
        this.shouldError = false;
    }

    getENS = async (_callerAddress) => {
        let promise = new Promise(async (resolve) => {
            await this.sleep();
            resolve(`${_callerAddress.slice(0,4)}...${_callerAddress.slice(-4)}`);

        });
        return promise;
    };

    getAllIssues = async (library) => {
        await sleep();

        return axios.get(`http://localhost:3030/getAllIssues`)
            .then(result => {
                return result.data;
            })
            .catch(error => {
                console.error(error);
            });
    };

    getIssueAddresses = async (library, issues) => {
        return axios.get(`http://localhost:3030/getIssueAddresses`)
            .then(result => {
                return result.data;
            })
            .catch(error => {
                console.error(error);
            });
    };

    getIssueIsOpen = async (library, issueId) => {
        return true;
    };

    getIssueIdFromAddress = async (library, address) => {
        return axios.get(`http://localhost:3030/getIssueIdFromAddress/${address}`)
            .then(result => {
                return result.data.issueId;
            })
            .catch(error => {
                console.error(error);
            });
    };

    getAddressById = async (library, externalUserId) => {
        return new Promise(async (resolve, reject) => {
            setTimeout(() => resolve("0xbcc58fb72409ba1cdec5dbcdd3cd6c42e3e04242"), 500);

        });


    };

    signMessage = async () => {
        const promise = new Promise(async (resolve, reject) => {
            resolve({});
        });
        return promise;
    };

    mintBounty = async (library, issueId, organization) => {
        const promise = new Promise(async (resolve, reject) => {
            await this.sleep();
            resolve({ "bountyAddress": "0x1abcD810374b2C0fCDD11cFA280Df9dA7970da4e", "txnReceipt": { events: ["0x1abcD810374b2C0fCDD11cFA280Df9dA7970da4e"] } });
        });
        return promise;
    };

    fundBounty = async (library, _bountyId, _tokenAddress, _value, _depositPeriodDays) => {
        const promise = new Promise(async (resolve, reject) => {
            await this.sleep(20);
            resolve({ events: [{ transactionHash: "0x1abcD810374b2C0fCDD11cFA280Df9dA7970da4e" }] });
        });
        return promise;
    };

    fundBountyWithNft = async (library, _bountyId, _tokenAddress, _value, _depositPeriodDays) => {
        const promise = new Promise(async (resolve, reject) => {
            await this.sleep(20);
            resolve({ events: [{ transactionHash: "0x1abcD810374b2C0fCDD11cFA280Df9dA7970da4e" }] });
        });
        return promise;
    };
    setFundingGoal = async (library, _bountyId, _fundingGoalToken, _fundingGoalVolume, _depositPeriodDays) => {
        const _tokenAddress = _fundingGoalToken.address;
        this.mockMutations.setFundingGoal(_bountyId, _fundingGoalVolume, _tokenAddress);
        const promise = new Promise(async (resolve, reject) => {
            await this.sleep(200);

            const volumeInWei = _fundingGoalVolume * 10 ** 18;
            const zeroAddress = "0x0000000000000000000000000000000000000000";
            resolve({ transactionHash: "0x1abcD810374b2C0fCDD11cFA280Df9dA7970da4e", events: [{ transactionHash: "0x1abcD810374b2C0fCDD11cFA280Df9dA7970da4e", args: [{}, _tokenAddress, volumeInWei] }] });
        });
        return promise;
    };
    setPayout = async (library, _bountyId, _payoutToken, _payoutVolume) => {
        const _tokenAddress = _payoutToken.address;
        this.mockMutations.setPayout(_bountyId, _payoutVolume, _tokenAddress);
        const promise = new Promise(async (resolve, reject) => {
            await this.sleep(200);

            const volumeInWei = _payoutVolume * 10 ** 18;
            const zeroAddress = "0x0000000000000000000000000000000000000000";
            resolve({ transactionHash: "0x1abcD810374b2C0fCDD11cFA280Df9dA7970da4e", events: [{ transactionHash: "0x1abcD810374b2C0fCDD11cFA280Df9dA7970da4e", args: [{}, _tokenAddress, volumeInWei] }] });
        });
        return promise;
            
    };
    setSupportingDocumentsComplete = async (library, _bountyId) => {
        this.mockMutations.setSupportingDocumentsComplete(_bountyId);
        const promise = new Promise(async (resolve, reject) => {
            await this.sleep(200);

            const zeroAddress = "0x0000000000000000000000000000000000000000";
            resolve({ transactionHash: "0x1abcD810374b2C0fCDD11cFA280Df9dA7970da4e", events: [{ transactionHash: "0x1abcD810374b2C0fCDD11cFA280Df9dA7970da4e", args: [{}] }] });
        });
        return promise;
            
    };


    setPayoutSchedule = async (library, _bountyId, _payoutSchedule) => {
        this.mockMutations.setTier(_bountyId, _payoutSchedule);
            
        const promise = new Promise(async (resolve, reject) => {
            await this.sleep(200);
            resolve({ transactionHash: "0x1abcD810374b2C0fCDD11cFA280Df9dA7970da4e", events: [{ transactionHash: "0x1abcD810374b2C0fCDD11cFA280Df9dA7970da4e", args: [_bountyId, _payoutSchedule] }] });
        });
        return promise;
        
    };

    setPayoutScheduleFixed = async (library, _bountyId, _payoutSchedule, _payoutToken) => {
            const promise = new Promise(async (resolve, reject) => {

            const tierVolumesInWei = _payoutSchedule.map((tier) => {
                const payoutVolumeInWei = tier * 10 ** _payoutToken.decimals;
                const payoutBigNumberVolumeInWei = ethers.BigNumber.from(
                    payoutVolumeInWei.toLocaleString('fullwide', {
                        useGrouping: false,
                    })
                );
                return payoutBigNumberVolumeInWei;
            });

            this.mockMutations.setTier(_bountyId, _payoutSchedule, _payoutToken.address);
            await this.sleep(200);
            resolve({ transactionHash: "0x1abcD810374b2C0fCDD11cFA280Df9dA7970da4e", events: [{ transactionHash: "0x1abcD810374b2C0fCDD11cFA280Df9dA7970da4e", args: [_bountyId, tierVolumesInWei, _payoutToken.address] }] });
        });
        return promise;
    };
    closeOngoing = async (library, _bountyId) => {
    
        const promise = new Promise(async (resolve, reject) => {
            await this.sleep(20);
        this.mockMutations.closeOngoing(_bountyId);
            resolve({ events: [{ transactionHash: "0x1abcD810374b2C0fCDD11cFA280Df9dA7970da4e" }] });
        });
        return promise;
    };




    balanceOf = async (library, _callerAddress, _tokenAddress) => {
        const promise = new Promise(async (resolve, reject) => {
            resolve(ethers.BigNumber.from("6600000000000000000"));
        });
        return promise;
    };

    allowance = async (library, _callerAddress, _tokenAddress, _bountyAddress) => {
        const promise = new Promise(async (resolve, reject) => {
            if (_tokenAddress === '0x5FbDB2315678afecb367f032d93F642f64180aa3' || _tokenAddress === '0x0000000000000000000000000000000000000000') {
                resolve(ethers.BigNumber.from("0"));
            } else {
                resolve(ethers.BigNumber.from("6600000000000000000"));
            }
        });
        return promise;
    };

    fetchNfts = async () => {
        const promise = new Promise((resolve, reject) => {
            axios.get('http://localhost:3030/fetchNfts')
                .then(result => {
                    resolve(result.data);
                })
                .catch(error => {
                    reject(error);
                });
        });

        return promise;
    };

    approve = async (library, _bountyAddress, _tokenAddress, _value) => {
        const promise = new Promise(async (resolve, reject) => {
            try {
                await this.sleep();
                if (this.shouldError) {
                    throw new Error();
                }
                resolve({});
            }
            catch (error) {
                reject({

                    title: 'Internal JSON',
                    message: 'Internal JSON-RPC error',
                    data: {
                        title: 'Internal JSON',
                        message: 'Internal JSON-RPC error'
                    }
                });
            }
        });
        return promise;
    };
    approveNFT = async (library, _bountyAddress, _tokenAddress, _value) => {
        const promise = new Promise(async (resolve, reject) => {
            try {
                await this.sleep();
                if (this.shouldError) {
                    throw new Error();
                }
                resolve({});
            }
            catch (error) {
                reject({

                    title: 'Internal JSON',
                    message: 'Internal JSON-RPC error',
                    data: {
                        title: 'Internal JSON',
                        message: 'Internal JSON-RPC error'
                    }
                });
            }
        });
        return promise;
    };

    async claimBounty(library, _bountyAddress /* _closer, _claimantAsset, tier*/) {
        return new Promise(async (resolve, reject) => {

            try {

                setTimeout(() => resolve({ transactionHash: _bountyAddress }), 2000);
            } catch (error) {
                reject(error);
            }
        });
    }


    userOwnedTokenBalances = async (library, _callerAddress, tokens) => {
        const promise = new Promise(async (resolve) => {
            const tokensInWallet = [];
            tokens.forEach(async (token) => {
                tokensInWallet.push(this.userBalanceForToken(library, token, _callerAddress));
            });
            resolve(Promise.all(tokensInWallet));
        });

        return promise;
    };

    userBalanceForToken = async (library, token, _callerAddress) => {
        let promise = new Promise(async (resolve) => {
            this.sleep();
            resolve(true);

        });

        return promise;
    };

    isWhitelisted = async (library, tokenAddress) => {
        const promise = new Promise(async (resolve, reject) => {
            try {
                resolve(true);
            } catch (err) {
                reject(err);
            }
        });
        return promise;
    };

    tokenAddressLimitReached = async (library, tokenAddress) => {
        const promise = new Promise(async (resolve, reject) => {
            try {
                resolve();
            } catch (err) {
                reject(err);
            }
        });
        return promise;
    };

    refundDeposit = async (library, _bountyId, _depositId) => {
        const promise = new Promise((resolve, reject) => {
            axios.get('http://localhost:3030/txnResponse')
                .then(async (result) => {
                    await this.sleep();
                    resolve(result.data);
                })
                .catch((error) => {
                    reject(error);
                });
        });
        return promise;
    };


    handleError(jsonRpcError, data) {
        let errorString = jsonRpcError?.data?.message;
        if (typeof jsonRpcError === 'string') {
            if (jsonRpcError.includes('Ambire user rejected the request')) { errorString = 'USER_DENIED_TRANSACTION'; }
            if (jsonRpcError.includes('Rejected Request')) { errorString = 'USER_DENIED_TRANSACTION'; }
        }

        if (jsonRpcError.message) {

            if (jsonRpcError.message.includes('Nonce too high.')) { errorString = 'NONCE_TO_HIGH'; }
            if (jsonRpcError.message.includes('User denied transaction signature')) { errorString = 'USER_DENIED_TRANSACTION'; }
            if (jsonRpcError.message.includes('MetaMask is having trouble connecting to the network')) { errorString = 'METAMASK_HAVING_TROUBLE'; }
            if (jsonRpcError.message.includes('Internal JSON-RPC error')) { errorString = 'INTERNAL_ERROR'; }
            if (jsonRpcError.message.includes('Set a higher gas fee')) { errorString = 'UNDERPRICED_TXN'; }
        }
        if (!errorString) {
            errorString = 'CALL_EXCEPTION';
        }
        for (const error of jsonRpcErrors) {
            const revertString = Object.keys(error)[0];
            if (errorString.includes(revertString)) {
                const title = error[revertString]['title'];
                const message = error[revertString].message(data);
                const link = error[revertString].link;
                const linkText = error[revertString].linkText;
                return { title, message, link, linkText };
            }
        }
        return 'Unknown Error';
    }
}

export default MockOpenQClient;