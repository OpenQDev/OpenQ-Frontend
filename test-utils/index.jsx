// Third party
import React from 'react';
// Cusotm
import { render, renderHook } from '@testing-library/react';
import { cleanup } from '@testing-library/react';

import StoreProvider from '../store/Store/TestStoreProvider';
import AuthProvider from '../store/AuthStore/TestAuthProvider';
import FundProvider from '../components/FundBounty/FundStore/TestFundProvider';
import TokenProvider from '../components/TokenSelection/TokenStore/TokenProvider';

// Add in any providers here if necessary:
// test-utils.js
// (ReduxProvider, ThemeProvider, etc)
afterEach(() => {
  cleanup();
});

const customRender = (ui, options = {}, storeProps, authProps) => {
  const IntersectionObserverMock = vi.fn(() => ({
    disconnect: vi.fn(),
    observe: vi.fn(),
    takeRecords: vi.fn(),
    unobserve: vi.fn(),
  }));
  vi.stubGlobal('IntersectionObserver', IntersectionObserverMock);
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <TokenProvider>{children}</TokenProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return render(ui, { wrapper: Providers, ...options });
};
const customHookRender = (hook, options = {}, storeProps, authProps, fundProps) => {
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <FundProvider StoreProps={fundProps}>
            <TokenProvider>{children}</TokenProvider>
          </FundProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return renderHook(hook, { wrapper: Providers, ...options });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, customHookRender as hookRender };
