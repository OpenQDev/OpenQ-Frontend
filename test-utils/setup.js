import { expect, beforeEach } from 'vitest';
import { cleanup } from '@testing-library/react';
import matchers from '@testing-library/jest-dom/matchers';

expect.extend(matchers);
beforeEach(() => {
  cleanup();
});
process.on('unhandledRejection', (reason) => {
  // eslint-disable-next-line no-console
  console.log(`FAILED TO HANDLE PROMISE REJECTION`);
  throw reason;
});
