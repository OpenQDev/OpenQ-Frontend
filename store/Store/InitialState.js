import Utils from '../../services/utils/Utils';

import OpenQClient from '../../services/ethers/OpenQClient';
import OpenQSubgraphClient from '../../services/subgraph/OpenQSubgraphClient';
import AuthService from '../../services/auth/AuthService';
import GithubRepository from '../../services/github/GithubRepository';
import TokenClient from '../../services/coins/TokenClient';
import Logger from '../../services/logger/Logger';
import OpenQPrismaClient from '../../services/openq-api/OpenQPrismaClient';
/* import SuperfluidClient from '../../services/SuperfluidClient/SuperfluidClient'; */

import MockGithubRepository from '../../services/github/MockGithubRepository';
import MockOpenQClient from '../../services/ethers/MockOpenQClient';
/* import MockOpenQSubgraphClient from '../../services/subgraph/MockOpenQSubgraphClient'; */
import MockTokenClient from '../../services/coins/MockTokenClient';
import MockOpenQPrismaClient from '../../services/openq-api/MockOpenQPrismaClient';
import MockAuthService from '../../services/auth/MockAuthService.js';

let InitialState = {};
switch (process.env.NEXT_PUBLIC_DEPLOY_ENV) {
  case 'local':
    InitialState = {
      accountData: {},
      openQClient: new MockOpenQClient(),
      authService: new MockAuthService(),
      githubRepository: new MockGithubRepository(),
      /*  openQSubgraphClient: new MockOpenQSubgraphClient(), */
      tokenClient: new MockTokenClient(),
      logger: new Logger('DEV'),
      utils: new Utils(),
      openQPrismaClient: new MockOpenQPrismaClient(),
      /*  superfluidClient: new SuperfluidClient(), */
    };
    break;
  case 'docker':
    InitialState = {
      accountData: {},
      openQClient: new OpenQClient(),
      authService: new AuthService(),
      githubRepository: new GithubRepository(),
      openQSubgraphClient: new OpenQSubgraphClient(),
      tokenClient: new TokenClient(),
      logger: new Logger('DEV'),
      utils: new Utils(),
      openQPrismaClient: new OpenQPrismaClient(),
      /*  superfluidClient: new SuperfluidClient(), */
    };
    break;
  case 'development':
    InitialState = {
      accountData: {},
      openQClient: new OpenQClient(),
      authService: new AuthService(),
      githubRepository: new GithubRepository(),
      openQSubgraphClient: new OpenQSubgraphClient(),
      tokenClient: new TokenClient(),
      logger: new Logger('DEV'),
      utils: new Utils(),
      openQPrismaClient: new OpenQPrismaClient(),
      /*  superfluidClient: new SuperfluidClient(), */
    };
    break;
  case 'staging':
    InitialState = {
      accountData: {},
      openQClient: new OpenQClient(),
      authService: new AuthService(),
      githubRepository: new GithubRepository(),
      openQSubgraphClient: new OpenQSubgraphClient(),
      tokenClient: new TokenClient(),
      logger: new Logger('DEV'),
      utils: new Utils(),
      openQPrismaClient: new OpenQPrismaClient(),
      /*  superfluidClient: new SuperfluidClient(), */
    };
    break;
  case 'production':
    InitialState = {
      accountData: {},
      openQClient: new OpenQClient(),
      authService: new AuthService(),
      githubRepository: new GithubRepository(),
      openQSubgraphClient: new OpenQSubgraphClient(),
      tokenClient: new TokenClient(),
      logger: new Logger('PROD'),
      utils: new Utils(),
      openQPrismaClient: new OpenQPrismaClient(),
      /*  superfluidClient: new SuperfluidClient(), */
    };
    break;
  default:
    throw Error('ENVIRONMENT NOT CONFIGURED CORRECTLY. Set an environment with DEPLOY_ENV');
}

export default InitialState;
