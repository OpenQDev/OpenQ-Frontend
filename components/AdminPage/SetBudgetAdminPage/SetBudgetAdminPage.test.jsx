/**
 * @vi-environment jsdom
 */
import React from 'react';
import { render, screen } from '../../../test-utils';
import SetBudgetAdminPage from '.';
import InitialState from '../../../store/Store/InitialState';
import Constants from '../../../test-utils/constant';
import userEvent from '@testing-library/user-event';
import MockOpenQClient from '../../../services/ethers/MockOpenQClient';

describe('SetBudgetAdminPage', () => {
  const bounty = Constants.bounty;
  beforeEach(() => {
    InitialState.openQClient.reset();
    const observe = vi.fn();
    const disconnect = vi.fn();
    window.IntersectionObserver = vi.fn(() => ({
      observe,
      disconnect,
    }));
  });
  describe('Split Price Contracts', () => {
    const splitBounty = Constants.bounty1;
    it('should allow user to update budget token and volume when no deposits', async () => {
      const noDepositSplitBounty = { ...splitBounty, deposits: [] };
      // ARRANGE
      const user = userEvent.setup();
      const setFundingGoal = vi.fn();
      const customInitialState = {
        ...InitialState,
        openQClient: new MockOpenQClient({ setFundingGoal }),
      };
      const { asFragment } = render(
        <SetBudgetAdminPage refreshBounty={() => {}} bounty={noDepositSplitBounty} />,
        {},
        customInitialState
      );
      expect(asFragment()).toMatchSnapshot('Snapshot: initial view');

      // ACT
      await user.type(screen.getByRole('textbox'), '100');
      await user.click(screen.getByRole('button', { name: 'select token' }));
      await user.click(screen.getByRole('button', { name: /link/i }));
      await user.click(await screen.findByRole('button', { name: 'Set New Budget' }));

      // ASSERT
      expect(await screen.findByText(/Updating Budget.../)).toBeInTheDocument();
      expect(screen.getByText(/our request is being processed.../)).toBeInTheDocument();
      const updatedTexts = await screen.findAllByText(/updated/i);
      expect(updatedTexts[0]).toBeInTheDocument();
      expect(await screen.findByText(/budget set to/i)).toBeInTheDocument();
      expect(await screen.findByText(/100.0 LINK/)).toBeInTheDocument();
      expect(setFundingGoal).toBeCalledWith(
        noDepositSplitBounty.bountyId,
        '100',
        '0x5FbDB2315678afecb367f032d93F642f64180aa3'
      );
    });
    it('should set payout token per default if payoutTokenVolume BUT allow user to update payout token and volume if no deposits', async () => {
      // payout token to DERC20, so showing that as default, but allowing change to Link since there are no deposits.
      // ARRANGE
      const noDepositButPayoutSplitBounty = {
        ...splitBounty,
        deposits: [],
        payoutTokenAddress: '0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0',
      };
      const user = userEvent.setup();
      const setFundingGoal = vi.fn();
      const customInitialState = {
        ...InitialState,
        openQClient: new MockOpenQClient({ setFundingGoal }),
      };
      const { asFragment } = render(
        <SetBudgetAdminPage refreshBounty={() => {}} bounty={noDepositButPayoutSplitBounty} />,
        {},
        customInitialState
      );
      expect(asFragment()).toMatchSnapshot('Snapshot: initial view');

      // ACT
      await user.type(screen.getByRole('textbox'), '100');
      const matic = screen.queryByText(/matic/i);
      expect(matic).not.toBeInTheDocument();
      const tokenButton = screen.getByRole('button', { name: 'select token' });
      await user.click(tokenButton);
      await user.click(screen.getByRole('button', { name: /link/i }));
      await user.click(await screen.findByRole('button', { name: 'Set New Budget' }));

      // ASSERT
      expect(await screen.findByText(/Updating Budget.../)).toBeInTheDocument();
      expect(screen.getByText(/our request is being processed.../)).toBeInTheDocument();
      const updatedTexts = await screen.findAllByText(/updated/i);
      expect(updatedTexts[0]).toBeInTheDocument();
      expect(await screen.findByText(/100.0 LINK/i)).toBeInTheDocument();
      expect(setFundingGoal).toBeCalledWith(
        noDepositButPayoutSplitBounty.bountyId,
        '100',
        '0x5FbDB2315678afecb367f032d93F642f64180aa3'
      );
    });
    it('should lock token but allow volume update when deposits', async () => {
      // ARRANGE
      const otherTokenDepositSplitBounty = { ...splitBounty, deposits: [Constants.deposit2] };
      const user = userEvent.setup();
      const setFundingGoal = vi.fn();
      const customInitialState = {
        ...InitialState,
        openQClient: new MockOpenQClient({ setFundingGoal }),
      };
      render(
        <SetBudgetAdminPage refreshBounty={() => {}} bounty={otherTokenDepositSplitBounty} />,
        {},
        customInitialState
      );

      // ACT
      await user.type(screen.getByRole('textbox'), '100');
      const selectToken = screen.getByRole('button', { name: 'select token' });
      expect(selectToken).toBeDisabled(true);
      expect(selectToken).toMatchSnapshot('Snapshot: button style changes to not appear as button');
      await user.click(await screen.findByRole('button', { name: 'Set New Budget' }));

      // ASSERT
      expect(await screen.findByText(/Updating Budget.../)).toBeInTheDocument();
      expect(screen.getByText(/our request is being processed.../)).toBeInTheDocument();
      const updatedTexts = await screen.findAllByText(/updated/i);
      expect(updatedTexts[0]).toBeInTheDocument();
      expect(await screen.findByText(/100.0 DERC20/i)).toBeInTheDocument();
      expect(setFundingGoal).toBeCalledWith(
        otherTokenDepositSplitBounty.bountyId,
        '100',
        otherTokenDepositSplitBounty.deposits[0].tokenAddress
      );
    });
  });
  describe('Fixed price contracts', () => {
    it('should allow user to update budget', async () => {
      // ARRANGE
      const user = userEvent.setup();
      const setFundingGoal = vi.fn();
      const customInitialState = {
        ...InitialState,
        openQClient: new MockOpenQClient({ setFundingGoal }),
      };
      render(<SetBudgetAdminPage refreshBounty={() => {}} bounty={bounty} />, {}, customInitialState);
      expect(screen.getByText('Set a New Budget for this Contract')).toBeInTheDocument();

      // ACT
      await user.type(screen.getByRole('textbox'), '100');
      await user.click(screen.getByRole('button', { name: 'select token' }));
      await user.click(screen.getByRole('button', { name: /link/i }));
      await user.click(await screen.findByRole('button', { name: 'Set New Budget' }));

      // ASSERT
      expect(await screen.findByText(/Updating Budget.../)).toBeInTheDocument();
      expect(screen.getByText(/our request is being processed.../)).toBeInTheDocument();
      const updatedTexts = await screen.findAllByText(/updated/i);
      expect(updatedTexts[0]).toBeInTheDocument();
      expect(await screen.findByText(/budget set to/i)).toBeInTheDocument();
      expect(await screen.findByText(/100.0 LINK/)).toBeInTheDocument();
      expect(setFundingGoal).toBeCalledWith(bounty.bountyId, '100', '0x5FbDB2315678afecb367f032d93F642f64180aa3');
    });
  });
});
