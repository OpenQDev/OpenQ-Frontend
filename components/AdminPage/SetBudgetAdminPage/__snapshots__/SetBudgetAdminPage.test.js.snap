// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetBudgetAdminPage Split Price Contracts should allow user to update budget token and volume when no deposits: Snapshot: initial view 1`] = `
<DocumentFragment>
   
  <div
    class="flex items-center gap-2"
  >
    Set a New Budget for this Contract
  </div>
  <div
    class="flex-1 items-center w-full mt-2"
  >
    <div
      class="flex space-x-4 w-full undefined"
    >
      <div
        class="flex w-full flex-row justify-between items-center px-4 input-field-big"
      >
        <div
          class=" bg-dark-mode"
        >
          <input
            aria-label="amount"
            autocomplete="off"
            class="font-semibold number outline-none bg-input-bg text-primary w-full"
            id="amount"
            placeholder="0.0"
            value=""
          />
        </div>
      </div>
      <div
        class="flex"
      >
        <button
          aria-label="select token"
          class="flex flex-row items-center p-0.5 px-2 btn-default"
        >
           
          <div
            class="flex h-4 w-4 items-center justify-center"
          >
            <img
              alt="selected token"
              class="rounded-full"
              data-nimg="1"
              decoding="async"
              height="40"
              loading="lazy"
              src="https://wallet-asset.matic.network/img/tokens/matic.svg"
              srcset="https://wallet-asset.matic.network/img/tokens/matic.svg 1x, https://wallet-asset.matic.network/img/tokens/matic.svg 2x"
              style="color: transparent;"
              width="40"
            />
          </div>
          <div
            class="flex pl-2 pr-1 text-primary"
          >
            MATIC
          </div>
          <div
            class="flex"
          >
            <svg
              class="h-4 w-4"
              fill="white"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
  </div>
  <div
    class="flex items-center"
  />
  <button
    class="btn-default"
    type="button"
  >
    Set New Budget
  </button>
</DocumentFragment>
`;

exports[`SetBudgetAdminPage Split Price Contracts should lock token but allow volume update when deposits: Snapshot: button style changes to not appear as button 1`] = `
<button
  aria-label="select token"
  class="flex flex-row items-center p-0.5 px-2 cursor-default"
  disabled=""
>
   
  <div
    class="flex h-4 w-4 items-center justify-center"
  >
    <img
      alt="selected token"
      class="rounded-full"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="https://wallet-asset.matic.network/img/tokens/dai.svg"
      srcset="https://wallet-asset.matic.network/img/tokens/dai.svg 1x, https://wallet-asset.matic.network/img/tokens/dai.svg 2x"
      style="color: transparent;"
      width="40"
    />
  </div>
  <div
    class="flex pl-2 pr-1 text-primary"
  >
    DERC20
  </div>
</button>
`;

exports[`SetBudgetAdminPage Split Price Contracts should set payout token per default if payoutTokenVolume BUT allow user to update payout token and volume if no deposits: Snapshot: initial view 1`] = `
<DocumentFragment>
   
  <div
    class="flex items-center gap-2"
  >
    Set a New Budget for this Contract
  </div>
  <div
    class="flex-1 items-center w-full mt-2"
  >
    <div
      class="flex space-x-4 w-full undefined"
    >
      <div
        class="flex w-full flex-row justify-between items-center px-4 input-field-big"
      >
        <div
          class=" bg-dark-mode"
        >
          <input
            aria-label="amount"
            autocomplete="off"
            class="font-semibold number outline-none bg-input-bg text-primary w-full"
            id="amount"
            placeholder="0.0"
            value=""
          />
        </div>
      </div>
      <div
        class="flex"
      >
        <button
          aria-label="select token"
          class="flex flex-row items-center p-0.5 px-2 btn-default"
        >
           
          <div
            class="flex h-4 w-4 items-center justify-center"
          >
            <img
              alt="selected token"
              class="rounded-full"
              data-nimg="1"
              decoding="async"
              height="40"
              loading="lazy"
              src="https://wallet-asset.matic.network/img/tokens/dai.svg"
              srcset="https://wallet-asset.matic.network/img/tokens/dai.svg 1x, https://wallet-asset.matic.network/img/tokens/dai.svg 2x"
              style="color: transparent;"
              width="40"
            />
          </div>
          <div
            class="flex pl-2 pr-1 text-primary"
          >
            DERC20
          </div>
          <div
            class="flex"
          >
            <svg
              class="h-4 w-4"
              fill="white"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
  </div>
  <div
    class="flex items-center"
  />
  <button
    class="btn-default"
    type="button"
  >
    Set New Budget
  </button>
</DocumentFragment>
`;
